python/lsst/pipe/tasks/calibrateImage.py

```diff
@@ -1501,3 +1640,293 @@ def _update_wcs_with_camera_model(self, exposure, cameraModel):
+            The exposure on which to compute the per-amp masked fraction.
+        detected_fraction : `float`
+            The current detected_fraction of the ``mask_planes`` for the
+            full detector.
+        mask_planes : `list`, `str`
+            The mask planes for which to evaluate the fraction.
+        bad_mask_planes : `list`, `str`
+            The mask planes to exclude from the computation.
+
+        Returns
+        -------
+        n_above_max_per_amp : `int`
+            The number of amplifiers with masked fractions above a maximum
+            value (set by the current full-detector ``detected_fraction``).
+        highest_detected_fraction_per_amp : `float`
+            The highest value of the per-amplifier fraction of masked pixels.
+        no_zero_det_amps : `bool`
+            A boolean representing whether any of the amplifiers has zero
+            masked pixels.
+        """
+        highest_detected_fraction_per_amp = -9.99
+        n_above_max_per_amp = 0
+        n_no_zero_det_amps = 0
+        no_zero_det_amps = True
+        amps = exposure.detector.getAmplifiers()
+        if amps is not None:
+            for ia, amp in enumerate(amps):
+                amp_bbox = amp.getBBox()
+                exp_bbox = exposure.getBBox()
+                if not exp_bbox.contains(amp_bbox):
+                    self.log.info("Bounding box of amplifier (%s) does not fit in exposure's "
+                                  "bounding box (%s).  Skipping...", amp_bbox, exp_bbox)
+                    continue
+                sub_image = exposure.subset(amp.getBBox())
+                detected_fraction_amp = self._compute_mask_fraction(sub_image.mask,
+                                                                    mask_planes,
+                                                                    bad_mask_planes)
+                self.log.debug("Current detected fraction for amplifier %s = %.3f",
+                               amp.getName(), detected_fraction_amp)
+                if detected_fraction_amp < 0.002:
+                    n_no_zero_det_amps += 1
+                    if n_no_zero_det_amps > 2:
+                        no_zero_det_amps = False
+                        break
+                highest_detected_fraction_per_amp = max(detected_fraction_amp,
+                                                        highest_detected_fraction_per_amp)
+                if highest_detected_fraction_per_amp > min(0.998, max(0.8, 3.0*detected_fraction)):
+                    n_above_max_per_amp += 1
+                    if n_above_max_per_amp > 2:
+                        break
+        else:
+            self.log.info("No amplifier object for detector %d, so skipping per-amp "
+                          "detection fraction checks.", exposure.detector.getId())
+        return n_above_max_per_amp, highest_detected_fraction_per_amp, no_zero_det_amps
+
+    def _remeasure_star_background(self, result):
```

Comment:
I put a lot of work into refactoring most of this method into a task in `meas_algorithms` on the u/jbosch/DM-48966 branch.  I don't think we want to do that on this ticket after all because of the implications for the configuration interface, but you might want to take a look at it and see if you want to instead want to apply some of the same ideas here now (entirely up to you):
 - I do think what you have here is too big for one method, and the handful I split it into on my branch seems like a decent split.
 - There's are a couple more loops below that would benefit from being rewritten as `for` loops that `break`, instead of `while` loops with a "are we still looping" boolean variable.